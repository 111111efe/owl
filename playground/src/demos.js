export const HELLO_WORLD = `class HelloWorld extends owl.core.Component {
  inlineTemplate = \`<div>Hello <t t-esc="props.name"/></div>\`;
}

const env = {
    qweb: new owl.core.QWeb()
};

const hello = new HelloWorld(env, { name: "World" });
hello.mount(document.body);
`;
