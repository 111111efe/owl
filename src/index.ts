import { QWeb } from "./qweb";
import { EventBus } from "./event_bus";
import { Component, PureComponent } from "./component";
import { Store, connect } from "./store";
import * as utils from "./utils";
export const core = {
  QWeb,
  EventBus,
  Component,
  PureComponent,
  Store,
  connect,
  utils
};
