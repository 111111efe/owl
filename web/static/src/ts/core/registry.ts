//------------------------------------------------------------------------------
// Registry code
//------------------------------------------------------------------------------
export class Registry<T> {
  private map: { [key: string]: T } = {};

  add(key: string, item: T): Registry<T> {
    if (key in this.map) {
      throw new Error(`Key ${key} already exists!`);
    }
    this.map[key] = item;
    return this;
  }

  get(key: string): T {
    return this.map[key];
  }
}

//------------------------------------------------------------------------------
// Main registry instance
//------------------------------------------------------------------------------
export const registry = new Registry();
