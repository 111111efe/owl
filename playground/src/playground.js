import { HELLO_WORLD } from "./demos.js";

class Playground extends owl.core.Component {
  inlineTemplate = `
    <div class="playground">
        <div class="left-bar">
            <button t-on-click="runCode">â–¶ Run</button>
            <div class="code-editor" t-ref="editor"></div>
        </div>
        <div class="content">
          <iframe t-ref="iframe"/>
        </div>
    </div>`;

  mounted() {
    this.editor = ace.edit(this.refs.editor);
    this.editor.session.setOption("useWorker", false);
    this.editor.setValue(HELLO_WORLD, -1);
    this.editor.setTheme("ace/theme/monokai");
    this.editor.session.setMode("ace/mode/javascript");
    const doc = this.refs.iframe.contentWindow.document;

    // add owl library
    const owlScript = doc.createElement("script");
    owlScript.src = "libs/owl.js";
    doc.body.appendChild(owlScript);
  }

  runCode() {
    const doc = this.refs.iframe.contentWindow.document;

    // // add owl library
    // const owlScript = doc.createElement("script");
    // owlScript.src = "libs/owl.js";
    // doc.body.appendChild(owlScript);

    const code = this.editor.getValue();
    const script = doc.createElement("script");
    script.type = "text/javascript";
    script.innerHTML = code;
    doc.body.appendChild(script);
    //  var thisDoc = this.contentWindow.document;
    //     if ( ! thisDoc.getElementById(scriptID)) {
    //         var scriptObj = thisDoc.createElement("script");
    //         scriptObj.type = "text/javascript";
    //         scriptObj.id = scriptId;
    //         scriptObj.innerHTML = script;
    //         thisDoc.body.appendChild(scriptObj);
    //     }
    // });
  }
}

document.addEventListener("DOMContentLoaded", async function() {
  const env = {
    qweb: new owl.core.QWeb()
  };

  const playground = new Playground(env);
  playground.mount(document.body);
});
