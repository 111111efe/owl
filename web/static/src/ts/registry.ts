import { Widget } from "./core/widget";
import { Env } from "./env";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------
interface Type<T> extends Function {
  new (...args: any[]): T;
}

type ActionWidget = Type<Widget<Env>>;

//------------------------------------------------------------------------------
// Registry code
//------------------------------------------------------------------------------
export class Registry {
  registries: { [key: string]: { [key: string]: any } } = {
    action: {}
  };

  add(type: "action", name: string, action: ActionWidget): Registry {
    if (name in this.registries[type]) {
      throw new Error(`Key ${name} already exists!`);
    }
    this.registries[type][name] = action;
    return this;
  }
}

//------------------------------------------------------------------------------
// Main registry instance
//------------------------------------------------------------------------------
export const registry = new Registry();
