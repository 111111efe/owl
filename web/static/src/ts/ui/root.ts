import { debounce } from "../core/utils";
import { Env } from "../env";
import { Controller } from "../store/action_manager_mixin";
import { State, Store } from "../store/store";
import { Widget } from "../widget";
import { HomeMenu } from "./home_menu";
import { Navbar } from "./navbar";
import { Notification } from "./notification";

//------------------------------------------------------------------------------
// Root Widget
//------------------------------------------------------------------------------

export class Root extends Widget<Store, State> {
  template = "web.web_client";
  widgets = { Navbar, HomeMenu, Notification };

  store: Store;

  constructor(env: Env, store: Store) {
    super(env, store);
    this.store = store;
    this.state = store.state;
  }

  mounted() {
    this.store.on("state_updated", this, this.updateState);
    this.store.on("rpc_status", this, this.toggleLoadingIndicator);
    this.store.on("update_action", this, this.applyController);
    if (this.store.lastController) {
      this.applyController(this.store.lastController);
    }

    // adding reactiveness to mobile/non mobile
    window.addEventListener("resize", <any>(
      debounce(this.applyMobileInterface.bind(this), 50)
    ));
  }

  applyMobileInterface() {
    const isMobile = window.innerWidth <= 768;
    if (isMobile !== this.env.isMobile) {
      this.env.isMobile = isMobile;
      this.render();
    }
  }

  toggleLoadingIndicator(status: "loading" | "notloading") {
    const method = status === "loading" ? "remove" : "add";
    (<any>this.refs.loading_indicator).classList[method]("d-none");
  }

  async applyController(controller: Controller) {
    const widget = await controller.create(this);
    if (widget) {
      // to do: call some public method of widget instead...
      (<HTMLElement>this.refs.content).appendChild(widget.el!);
      widget.__mount();
      widget.el!.classList.add("o_action_controller");
      this.store.activateController(controller);
    }
  }
}
