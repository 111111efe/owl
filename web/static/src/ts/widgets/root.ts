import { INotification } from "../core/notifications";
import { Widget } from "./widget";
import { Env } from "../env";
import { MenuInfo, MenuItem, getAppAndAction } from "../misc/menu_helpers";
import { ActionStack } from "../services/action_manager";
import { ActionContainer } from "./action_container";
import { HomeMenu } from "./home_menu";
import { Navbar } from "./navbar";
import { Notification } from "./notification";
import { Query } from "../core/router";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

interface State {
  notifications: INotification[];
  stack: ActionStack;
  inHome: boolean;
  currentApp: MenuItem | null;
}

interface Props {
  menuInfo: MenuInfo;
}

//------------------------------------------------------------------------------
// Root Widget
//------------------------------------------------------------------------------

export class Root extends Widget<Props, State> {
  template = "web.web_client";
  widgets = { Navbar, Notification, HomeMenu, ActionContainer };

  state: State = {
    notifications: [],
    stack: [],
    inHome: false,
    currentApp: null
  };

  constructor(env: Env, props: Props) {
    super(env, props);
    const query = this.env.router.getQuery();
    let { app, actionId } = getAppAndAction(props.menuInfo, query);
    this.state.currentApp = app;
    if (!actionId) {
      this.state.inHome = true;
    }
  }
  mounted() {
    // notifications
    this.env.notifications.on("notifications_updated", this, notifs =>
      this.updateState({ notifications: notifs })
    );

    // actions
    this.env.actionManager.on("action_stack_updated", this, stack =>
      this.updateState({ stack, inHome: false })
    );
    this.env.router.on("query_changed", this, this.updateAction);
    this.updateAction(this.env.router.getQuery());
  }

  private updateAction(query: Query) {
    let { app, actionId } = getAppAndAction(this.props.menuInfo, query);
    this.updateAppState(app, actionId);
  }

  private updateAppState(app: MenuItem | null, actionId: number | null) {
    const newApp = app || this.state.currentApp;
    if (actionId) {
      const query: Query = { action_id: String(actionId) };
      const menuId = newApp ? newApp.menuId : false;
      if (menuId) {
        query.menu_id = String(menuId);
      }
      if (app) {
        this.updateState({ currentApp: app });
      }
      this.env.router.navigate(query);
      this.env.actionManager.doAction(actionId);
    } else {
      this.updateState({ inHome: true, currentApp: newApp });
    }
  }

  toggleHome() {
    this.updateState({ inHome: !this.state.inHome });
  }

  openMenu(menu: MenuItem) {
    const app = this.props.menuInfo.menuMap[menu.menuId]!;
    this.updateAppState(app, menu.actionId);
  }
}
