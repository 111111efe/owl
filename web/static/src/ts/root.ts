import { Widget } from "./core/widget";
import { Navbar } from "./widgets/navbar";
import { ActionWidget } from "./services/action_manager";
import { INotification } from "./services/notifications";
import { Notification } from "./widgets/notification";
import { Env } from "./env";

const template = `
    <div class="o_web_client">
        <t t-widget="Navbar"/>
        <div class="o_content" t-ref="content"/>
        <div class="o_notification_container">
          <t t-foreach="state.notifications" t-as="notif">
            <t t-widget="Notification" t-props="notif"/>
          </t>
        </div>
    </div>
`;

export class Root extends Widget<Env, {}> {
  name = "root";
  template = template;
  widgets = { Navbar, Notification };
  content: Widget<Env, {}> | null = null;

  state: { notifications: INotification[] } = { notifications: [] };

  mounted() {
    this.env.actionManager.on("action_ready", this, this.setContentWidget);
    this.env.notifications.on("notification_added", this, this.addNotif);
    this.env.notifications.on("notification_removed", this, this.removeNotif);

    const actionWidget = this.env.actionManager.getCurrentAction();
    if (actionWidget) {
      this.setContentWidget(actionWidget);
    }
  }

  async setContentWidget(actionWidget: ActionWidget) {
    const currentWidget = this.content;
    const newWidget = new actionWidget.Widget(this, actionWidget.props);
    await newWidget.mount(<HTMLElement>this.refs.content);
    if (currentWidget) {
      currentWidget.destroy();
    }
    this.content = newWidget;
  }

  addNotif(notif: INotification) {
    const notifications = this.state.notifications.concat(notif);
    this.updateState({ notifications });
  }
  removeNotif(notif: INotification) {
    const notifs = this.state.notifications.filter(f => f.id !== notif.id);
    this.updateState({ notifications: notifs });
  }
}
