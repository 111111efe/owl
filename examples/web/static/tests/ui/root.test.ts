import { Registry } from "../../../../../src/registry";
import { Root } from "../../src/ts/ui/root";
import * as helpers from "../helpers";
import { makeTestEnv } from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;

beforeEach(() => {
  fixture = helpers.makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

test("can be rendered (in home menu)", async () => {
  const testEnv = makeTestEnv();
  const root = new Root(testEnv, testEnv.store);
  await root.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("if url has action_id, will render action and navigate to proper menu_id", async () => {
  const router = new helpers.MockRouter({ action_id: "131" });
  const testEnv = makeTestEnv({ services: <any>{ router } });
  await helpers.nextTick();

  const root = new Root(testEnv, testEnv.store);
  await root.mount(fixture);
  await helpers.nextTick();
  expect(router.getQuery()).toEqual({
    action_id: "131",
    menu_id: "96"
  });
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("start with no action => clicks on client action => discuss is rendered", async () => {
  const testEnv = makeTestEnv();
  const root = new Root(testEnv, testEnv.store);
  await root.mount(fixture);

  expect(testEnv.services.router.getQuery()).toEqual({ home: true });

  // discuss menu item
  await (<any>document.querySelector('[data-menu="96"]')).click();
  await helpers.nextTick();
  expect(fixture.innerHTML).toMatchSnapshot();
  expect(testEnv.services.router.getQuery()).toEqual({
    action_id: "131",
    menu_id: "96"
  });
});

test("clicks on client action with invalid key => empty widget is rendered + warning", async () => {
  const testEnv = makeTestEnv({ actionRegistry: new Registry() });
  const root = new Root(testEnv, testEnv.store);
  await root.mount(fixture);

  // discuss menu item
  await (<any>document.querySelector('[data-menu="96"]')).click();
  await helpers.nextTick();

  expect(fixture.innerHTML).toMatchSnapshot();
});

test("open act window action with invalid viewtype => empty widget is rendered + warning", async () => {
  const testEnv = makeTestEnv({ viewRegistry: new Registry() });
  const root = new Root(testEnv, testEnv.store);
  await root.mount(fixture);

  // note menu item
  await (<any>document.querySelector('[data-menu="205"]')).click();
  await helpers.nextTick();

  expect(fixture.innerHTML).toMatchSnapshot();
});
