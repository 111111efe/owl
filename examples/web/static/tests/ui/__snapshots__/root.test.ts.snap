// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered (in home menu) 1`] = `"<div class=\\"o_web_client\\"><div class=\\"o_navbar o_in_home\\"></div><div class=\\"o_home_menu\\"><div class=\\"o_apps\\"><a href=\\"#menu_id=96&amp;action_id=131\\" class=\\"o_app\\" data-menu=\\"96\\"><i class=\\"fa fa-comment o_app_icon fa-3x fa-fw\\"></i><div class=\\"o_caption\\">Discuss</div></a><a href=\\"#menu_id=205&amp;action_id=250\\" class=\\"o_app\\" data-menu=\\"205\\"><i class=\\"fa fa-pen o_app_icon fa-3x fa-fw\\"></i><div class=\\"o_caption\\">Notes</div></a><a href=\\"#menu_id=409&amp;action_id=597\\" class=\\"o_app\\" data-menu=\\"409\\"><i class=\\"fa fa-handshake o_app_icon fa-3x fa-fw\\"></i><div class=\\"o_caption\\">CRM</div></a></div></div><div class=\\"o_content o_hidden\\"></div><div class=\\"o_notification_container\\"></div><div class=\\"o_loading d-none\\">Loading</div></div>"`;

exports[`clicks on client action with invalid key => empty widget is rendered + warning 1`] = `"<div class=\\"o_web_client\\"><div class=\\"o_navbar\\"><a aria-label=\\"Applications\\" class=\\"o_title fa fa-th\\" href=\\"#\\" title=\\"Applications\\" accesskey=\\"h\\"></a><a class=\\"o_menu_brand\\" href=\\"\\" role=\\"button\\">Discuss</a><ul class=\\"o_menu_sections\\"></ul></div><div class=\\"o_content\\"><div class=\\"o_action_controller\\"></div></div><div class=\\"o_notification_container\\"><div class=\\"o_notification o_error\\"><div class=\\"o_notification_title\\"><span class=\\"o_icon fa fa-3x fa-exclamation\\" role=\\"img\\" aria-label=\\"Notification Invalid Client Action\\" title=\\"Notification Invalid Client Action\\"></span>Invalid Client Action</div><div class=\\"o_notification_content\\">Cannot find widget 'mail.discuss' in the action registry</div></div></div><div class=\\"o_loading d-none\\">Loading</div></div>"`;

exports[`if url has action_id, will render action and navigate to proper menu_id 1`] = `"<div class=\\"o_web_client\\"><div class=\\"o_navbar\\"><a aria-label=\\"Applications\\" class=\\"o_title fa fa-th\\" href=\\"#\\" title=\\"Applications\\" accesskey=\\"h\\"></a><a class=\\"o_menu_brand\\" href=\\"\\" role=\\"button\\">Discuss</a><ul class=\\"o_menu_sections\\"></ul></div><div class=\\"o_content\\"><div class=\\"o_discuss o_action_controller\\"><div class=\\"o_control_panel\\"><ol class=\\"breadcrumb\\" role=\\"navigation\\"><li class=\\"breadcrumb-item active\\">Discuss</li></ol><div class=\\"o_cp_searchview\\" role=\\"search\\"><div class=\\"o_searchview\\" role=\\"search\\" aria-autocomplete=\\"list\\"><span class=\\"o_searchview_more fa\\" title=\\"Advanced Search...\\" role=\\"img\\" aria-label=\\"Advanced Search...\\"></span></div></div><div class=\\"o_cp_left\\"><div class=\\"o_cp_buttons\\" role=\\"toolbar\\" aria-label=\\"Control panel toolbar\\"></div><aside class=\\"o_cp_sidebar\\"></aside></div><div class=\\"o_cp_right\\"><div class=\\"btn-group o_search_options\\" role=\\"search\\"></div><nav class=\\"o_cp_pager\\" role=\\"search\\" aria-label=\\"Pager\\"></nav><nav class=\\"btn-group o_cp_switch_buttons\\" role=\\"toolbar\\" aria-label=\\"View switcher\\"></nav></div></div><div class=\\"o_content\\"><div class=\\"o_discuss_sidebar\\"><span>DISCUSS!!</span></div><div class=\\"o_discuss_content\\"><button>Reset first counter</button><button>Reset counter 2 in 3s</button><button>Toggle Clock/counters</button><button>Toggle Color</button><button>Rerender this widget</button><input><div><button>-</button><span style=\\"font-weight:bold\\">Value: 4</span><button>+</button></div><div><button>-</button><span style=\\"font-weight:bold\\">Value: 400</span><button>+</button></div><div><span>Current Color: </span>red</div><button>Add notif</button><button>Add sticky notif</button><button>Add warning</button></div></div></div></div><div class=\\"o_notification_container\\"></div><div class=\\"o_loading d-none\\">Loading</div></div>"`;

exports[`open act window action with invalid viewtype => empty widget is rendered + warning 1`] = `"<div class=\\"o_web_client\\"><div class=\\"o_navbar\\"><a aria-label=\\"Applications\\" class=\\"o_title fa fa-th\\" href=\\"#\\" title=\\"Applications\\" accesskey=\\"h\\"></a><a class=\\"o_menu_brand\\" href=\\"\\" role=\\"button\\">Notes</a><ul class=\\"o_menu_sections\\"></ul></div><div class=\\"o_content\\"><div class=\\"o_action_controller\\"></div></div><div class=\\"o_notification_container\\"><div class=\\"o_notification o_error\\"><div class=\\"o_notification_title\\"><span class=\\"o_icon fa fa-3x fa-exclamation\\" role=\\"img\\" aria-label=\\"Notification Invalid View type\\" title=\\"Notification Invalid View type\\"></span>Invalid View type</div><div class=\\"o_notification_content\\">Cannot find view of type 'kanban' in the view registry</div></div></div><div class=\\"o_loading d-none\\">Loading</div></div>"`;

exports[`start with no action => clicks on client action => discuss is rendered 1`] = `"<div class=\\"o_web_client\\"><div class=\\"o_navbar\\"><a aria-label=\\"Applications\\" class=\\"o_title fa fa-th\\" href=\\"#\\" title=\\"Applications\\" accesskey=\\"h\\"></a><a class=\\"o_menu_brand\\" href=\\"\\" role=\\"button\\">Discuss</a><ul class=\\"o_menu_sections\\"></ul></div><div class=\\"o_content\\"><div class=\\"o_discuss o_action_controller\\"><div class=\\"o_control_panel\\"><ol class=\\"breadcrumb\\" role=\\"navigation\\"><li class=\\"breadcrumb-item active\\">Discuss</li></ol><div class=\\"o_cp_searchview\\" role=\\"search\\"><div class=\\"o_searchview\\" role=\\"search\\" aria-autocomplete=\\"list\\"><span class=\\"o_searchview_more fa\\" title=\\"Advanced Search...\\" role=\\"img\\" aria-label=\\"Advanced Search...\\"></span></div></div><div class=\\"o_cp_left\\"><div class=\\"o_cp_buttons\\" role=\\"toolbar\\" aria-label=\\"Control panel toolbar\\"></div><aside class=\\"o_cp_sidebar\\"></aside></div><div class=\\"o_cp_right\\"><div class=\\"btn-group o_search_options\\" role=\\"search\\"></div><nav class=\\"o_cp_pager\\" role=\\"search\\" aria-label=\\"Pager\\"></nav><nav class=\\"btn-group o_cp_switch_buttons\\" role=\\"toolbar\\" aria-label=\\"View switcher\\"></nav></div></div><div class=\\"o_content\\"><div class=\\"o_discuss_sidebar\\"><span>DISCUSS!!</span></div><div class=\\"o_discuss_content\\"><button>Reset first counter</button><button>Reset counter 2 in 3s</button><button>Toggle Clock/counters</button><button>Toggle Color</button><button>Rerender this widget</button><input><div><button>-</button><span style=\\"font-weight:bold\\">Value: 4</span><button>+</button></div><div><button>-</button><span style=\\"font-weight:bold\\">Value: 400</span><button>+</button></div><div><span>Current Color: </span>red</div><button>Add notif</button><button>Add sticky notif</button><button>Add warning</button></div></div></div></div><div class=\\"o_notification_container\\"></div><div class=\\"o_loading d-none\\">Loading</div></div>"`;
